torch/jit/__init__.py

```diff
@@ -671,38 +694,53 @@ def f(x):
+        for method_name, example_inputs in inputs.items():
+            methods_inputs[getattr(mod, method_name)] = (method_name, example_inputs)
+    else:
+        methods_inputs[mod] = (getattr(mod, '__name__', 'forward'), inputs)
+        if hasattr(mod, '__self__') and isinstance(mod.__self__, torch.nn.Module):
+            mod = mod.__self__
+
     executor_options = {'optimize': bool(optimize)}
-    # Special case for common case of passing a single Tensor
-    if isinstance(example_inputs, (torch.Tensor, dict)):
-        example_inputs = (example_inputs,)
-    # done primarily so that weird iterables fail here and not pybind11 code
-    elif not isinstance(example_inputs, tuple):
-        example_inputs = tuple(example_inputs)
-    var_lookup_fn = _create_interpreter_name_lookup_fn(0)
 
-    if isinstance(func, torch.nn.Module):
+    module = None
+    if isinstance(mod, torch.nn.Module):
         if _module_class is None:
             _module_class = TopLevelTracedModule
-        traced = _module_class(func, **executor_options)
-        traced._c._create_method_from_trace('forward', func, example_inputs,
-                                            var_lookup_fn, _force_outplace)
-    else:
-        name = getattr(func, '__name__', 'forward')
-        if name == '<lambda>':
-            name = '_lambda'  # make name a valid identifier
-        traced = torch._C._create_function_from_trace(name, func, example_inputs,
-                                                      var_lookup_fn,
-                                                      _force_outplace)
-
-    # Check the trace against new traces created from user-specified inputs
-    if check_trace:
-        if check_inputs is not None:
-            _check_trace(check_inputs, func, executor_options, traced, check_tolerance, _force_outplace)
+        module = _module_class(mod, **executor_options)
+
+    var_lookup_fn = _create_interpreter_name_lookup_fn(0)
+
+    for func, name_inputs in methods_inputs.items():
+        method_name = name_inputs[0]
+        example_inputs = name_inputs[1]
+        if isinstance(example_inputs, (torch.Tensor, dict)):
+            example_inputs = (example_inputs,)
+        # done primarily so that weird iterables fail here and not pybind11 code
+        elif not isinstance(example_inputs, tuple):
+            example_inputs = tuple(example_inputs)
+
+        if module:
+            module._c._create_method_from_trace(method_name, func, example_inputs, var_lookup_fn, _force_outplace)
+            check_trace_method = module._c._get_method(method_name)
         else:
-            _check_trace([example_inputs], func, executor_options, traced, check_tolerance, _force_outplace)
+            if method_name == '<lambda>':
+                method_name = '_lambda'  # make name a valid identifier
+            traced = torch._C._create_function_from_trace(method_name, func, example_inputs, var_lookup_fn, _force_outplace)
+            module = traced
```

Comment:
What is going on here? The method_inputs.items() loop is only sensible if the thing is a module. I feel like this could be refactored to be a lot more readable.